ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,332 @@
+          // 1. remove heartbeats older than window
+          while (!heartbeats.isEmpty()) {
+            Entry<Long, HeartbeatInfo> nestedEntry = heartbeats.firstEntry();
+            long timestamp = nestedEntry.getKey();
+            if (timestamp < windowStart) {
+              heartbeats.pollFirstEntry();
+            } else {
+              break;
+            }
+          }
+          // 2. count consecutive missed heartbeats and mark as up
+          downPolicy(host, windowStart, windowEnd, heartbeats);
+        }
+      }
+    }
+
+
+    private void downPolicy(final HostInfo host, final long windowStart, final long windowEnd,
+                            final TreeMap<Long, HeartbeatInfo> heartbeats) {
+      long missedCount = 0;
+      long prev = windowStart;
+      // No heartbeat received in this window
+      if (heartbeats.isEmpty()) {
+        aliveHosts.put(host.toString(), false);
+        return;
+      }
+      for (long ts : heartbeats.keySet()) {
+
+        //Don't count heartbeats after window end
+        if (ts >= windowEnd) {
+          continue;
+        }
+        if (ts - configs.heartbeatSendIntervalMs > prev) {
+          missedCount = (ts - prev - 1) / configs.heartbeatSendIntervalMs;
+        } else {
+          missedCount = 0;
+        }
+        prev = ts;
+      }
+      // Check frame from last received heartbeat to window end
+      if (windowEnd - prev - 1 > 0) {
+        missedCount = (windowEnd - prev - 1) / configs.heartbeatSendIntervalMs;
+      }
+
+      if (missedCount < configs.heartbeatMissedThreshold) {
+        aliveHosts.put(host.toString() , true);
+
+      } else {
+        aliveHosts.put(host.toString(), false);
+      }
+    }
+  }
+
+  class SendHeartbeatTask implements Runnable {
+
+    @Override
+    public void run() {
+      //broadcast heartbeat
+      remoteHostsURI.forEach((uri, bool) -> serviceContext.getKsqlClient().makeHeartbeatRequest(
+          uri, localHostInfo, System.currentTimeMillis()));
+    }
+  }
+
+  class DiscoverClusterTask implements Runnable {
+
+    @Override
+    public void run() {
+
+      List<PersistentQueryMetadata> currentQueries = engine.getPersistentQueries();
+      if (currentQueries.isEmpty()) {
+        return;
+      }
+
+      Set<HostInfo> uniqueHosts = currentQueries.stream()
+          .map(queryMetadata -> ((QueryMetadata)queryMetadata).getKafkaStreams().allMetadata())
+          .flatMap(Collection::stream)
+          .map(StreamsMetadata::hostInfo)
+          .filter(hostInfo -> !(hostInfo.host().equals(localHostInfo.host())
+              && hostInfo.port() == (localHostInfo.port())))
+          .collect(Collectors.toSet());
+
+      for (HostInfo hostInfo : uniqueHosts) {
+        // First time discovered
+        if (remoteHostsInfo.putIfAbsent(hostInfo, true) == null) {
+          URI remoteUri = buildLocation(localURL, hostInfo);
+          remoteHostsURI.put(remoteUri, true);
+          aliveHosts.put(hostInfo.toString(), true);
```

Comment:
should w consider it alive before it gets heartbeats? Maybe it's best to start off with `false` here

Reply:
This is a design decision we made to be optimistic when new servers are added to the cluster. We will forward queries and the worst that can happen is that they error out. The opposite would be to not serve queries which we don't want.

Reply:
makes sense - please document this inline as well